pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +955,364 @@ class StateVectorCudaMPI
+
+            if (localCSRMatrix.values.size() != 0) {
+                d_tmp.zeroInit();
+
+                DataBuffer<index_type, int> d_csrOffsets{
+                    localCSRMatrix.csrOffsets.size(), device_id, stream_id,
+                    true};
+                DataBuffer<index_type, int> d_columns{
+                    localCSRMatrix.columns.size(), device_id, stream_id, true};
+                DataBuffer<CFP_t, int> d_values{localCSRMatrix.values.size(),
+                                                device_id, stream_id, true};
+
+                d_csrOffsets.CopyHostDataToGpu(localCSRMatrix.csrOffsets.data(),
+                                               localCSRMatrix.csrOffsets.size(),
+                                               false);
+                d_columns.CopyHostDataToGpu(localCSRMatrix.columns.data(),
+                                            localCSRMatrix.columns.size(),
+                                            false);
+                d_values.CopyHostDataToGpu(localCSRMatrix.values.data(),
+                                           localCSRMatrix.values.size(), false);
+
+                // CUSPARSE APIs
+                cusparseSpMatDescr_t mat;
+                cusparseDnVecDescr_t vecX, vecY;
+
+                size_t bufferSize = 0;
+                cusparseHandle_t handle = getCusparseHandle();
+
+                // Create sparse matrix A in CSR format
+                PL_CUSPARSE_IS_SUCCESS(cusparseCreateCsr(
+                    /* cusparseSpMatDescr_t* */ &mat,
+                    /* int64_t */ num_rows_local,
+                    /* int64_t */ num_cols_local,
+                    /* int64_t */ nnz_local,
+                    /* void* */ d_csrOffsets.getData(),
+                    /* void* */ d_columns.getData(),
+                    /* void* */ d_values.getData(),
+                    /* cusparseIndexType_t */ compute_type,
+                    /* cusparseIndexType_t */ compute_type,
+                    /* cusparseIndexBase_t */ CUSPARSE_INDEX_BASE_ZERO,
+                    /* cudaDataType */ data_type));
+
+                // Create dense vector X
+                PL_CUSPARSE_IS_SUCCESS(cusparseCreateDnVec(
+                    /* cusparseDnVecDescr_t* */ &vecX,
+                    /* int64_t */ num_cols_local,
+                    /* void* */ BaseType::getData(),
+                    /* cudaDataType */ data_type));
+
+                // Create dense vector y
+                PL_CUSPARSE_IS_SUCCESS(cusparseCreateDnVec(
+                    /* cusparseDnVecDescr_t* */ &vecY,
+                    /* int64_t */ num_rows_local,
+                    /* void* */ d_tmp.getData(),
+                    /* cudaDataType */ data_type));
+
+                // allocate an external buffer if needed
+                PL_CUSPARSE_IS_SUCCESS(cusparseSpMV_bufferSize(
+                    /* cusparseHandle_t */ handle,
+                    /* cusparseOperation_t */ CUSPARSE_OPERATION_NON_TRANSPOSE,
```

Comment:
Should we move the CONSTANTS with the other constants like `data_type` at the top of the function (and maybe define `operation_type` and `algo_type`, say)?

Reply:
done!